runtime/lua/vim/diagnostic.lua

```diff
@@ -1543,6 +1548,46 @@ M.handlers.underline = {
   end,
 }
 
+---@param bufnr integer
+---@param ns integer
+---@param diagnostics table<integer, vim.Diagnostic[]>
+---@param opts vim.diagnostic.Opts.VirtualText
+local function render_virtual_text(bufnr, ns, diagnostics, opts)
+  api.nvim_buf_clear_namespace(bufnr, ns, 0, -1)
```

Comment:
I'm not sure if this approach (which is the same as what I'm doing in #31959) is compatible with the diagnostic caching in this module.